File: Parcial-I/Pregunta-III/buddysystem.py
import sys

def is_power_of_two(n):
    return (n & (n - 1) == 0) and n != 0

def get_power_of_two(n):
    if is_power_of_two(n):
        return n
    else:
        while not is_power_of_two(n):
            n += 1
    return n

class Node:
    def __init__(self, size, name, interval, parent, is_free=True):
        self.size = size
        self.name = name
        self.left = None
        self.right = None
        self.parent = parent

        self.start_interval = interval[0]
        self.end_interval = interval[1]

        self.is_free = is_free

class BuddySystem:
    reserved = []

    def __init__(self, total_memory):
        value = get_power_of_two(total_memory)
        self.root = Node(value, None, (0, value - 1), None)

    def add_process(self, size, name):
        if name in [i.name for i in self.reserved]:
            print(f"Error: El proceso {name} ya se encuentra reservado.")
        else:
            self.allocate_space(self.root, size, name)

    '''def allocate_space(self, node, size, name):
        if node.size == size and node.is_free:
            node.name = name
            node.is_free = False
            self.mark_parents_as_not_free(node.parent)
            self.reserved.append(node)
        elif node.size == size and not node.is_free:
            print(f"No hay suficiente memoria para el proceso {name} de tamaño {size}")
        elif node.size > size:
            if (node.size // 2) < size: # Cuando un proceso no es potencia de 2
                node.name = name
                node.is_free = False
                self.reserved.append(node)         
            elif (node.size // 2) >= size and not node.left and not node.right: # Cuando un proceso no tiene hijos
                node.left = Node(node.size // 2, None, (node.start_interval, node.start_interval + (node.size//2) - 1), node)
                node.right = Node(node.size // 2, None, (node.start_interval + node.size//2, node.end_interval), node)
                self.allocate_space(node.left, size, name)
            elif (node.size // 2) >= size and node.left and node.right: # Cuando un proceso tiene hijos
                left_free = self.find_free_child(node.left, size)
                right_free = self.find_free_child(node.right, size)

                if left_free is not None: # Ver si hay espacio en el hijo izq. o sus hijos
                    self.allocate_space(node.left, size, name)
                elif right_free is not None: # Ver si hay espacio en el hijo izq. o sus hijos
                    self.allocate_space(node.right, size, name)
                else:
                    print(f"No hay suficiente memoria para el proceso {name} de tamaño {size}")
        else:
            print(f"No hay suficiente memoria para el proceso {name} de tamaño {size}")'''
    
    def allocate_space(self, node, size, name):
        if node.size == size and node.is_free:
            node.name = name
            node.is_free = False
            self.mark_parents_as_not_free(node.parent)
            self.reserved.append(node)
        elif node.size == size and not node.is_free:
            print(f"No hay suficiente memoria para el proceso {name} de tamaño {size}")
        elif node.size > size:
            # 1. Cuando un proceso no es potencia de 2
            if (node.size // 2) < size:
                node.name = name
                node.is_free = False
                self.reserved.append(node)
            # 2. Cuando un proceso es potencia de 2 y no tiene hijos
            elif (node.size // 2) >= size and not node.left and not node.right:
                # 2.1 Cuando el proceso que se está viendo es la raiz
                if node == self.root and not node.is_free:
                    print(f"No hay suficiente memoria para el proceso {name} de tamaño {size}")
                    return
                node.left = Node(node.size // 2, None, (node.start_interval, node.start_interval + (node.size//2) - 1), node)
                node.right = Node(node.size // 2, None, (node.start_interval + node.size//2, node.end_interval), node)
                self.allocate_space(node.left, size, name)
            # 3. Cuando un proceso es potencia de 2 y tiene hijos
            elif (node.size // 2) >= size and node.left and node.right:
                left_free = self.find_free_child(node.left, size)
                right_free = self.find_free_child(node.right, size)

                # 3.1 Ver si hay espacio en el hijo izq. o sus hijos
                if left_free is not None:
                    self.allocate_space(node.left, size, name)
                # 3.2 Ver si hay espacio en el hijo izq. o sus hijos
                elif right_free is not None:
                    self.allocate_space(node.right, size, name)
                else:
                    print(f"No hay suficiente memoria para el proceso {name} de tamaño {size}")
        else:
            print(f"No hay suficiente memoria para el proceso {name} de tamaño {size}")

    def deallocate_process(self, name):
        node = self.find_node(self.root, name)
        if node is None:
            print(f"El proceso {name} no existe o no tiene memoria reservada.")
        else:
            node.name = None
            node.is_free = True
            if node.parent.right.is_free:
                node.parent.left = None
                node.parent.right = None
                node.parent.is_free = True

    ### Auxiliar Functions ####

    def mark_parents_as_not_free(self, node):
        if node:
            node.is_free = False
            self.mark_parents_as_not_free(node.parent)

    def find_free_child(self, node, needed_space):
        if node:
            if node.is_free and node.size >= needed_space:
                return node.size
            else:
                result = self.find_free_child(node.left, needed_space)
                if not result:
                    result = self.find_free_child(node.right, needed_space)
                return result
        else:
            return None
        
    def find_node(self, node, name):
        if node:
            if node.name == name:
                return node
            else:
                result = self.find_node(node.left, name)
                if not result:
                    result = self.find_node(node.right, name)
                return result
        else:
            return None

    def print_not_reserved(self, node):
        if node:
            self.print_not_reserved(node.right)
            if node.is_free:
                print("(", node.start_interval,"->", node.end_interval,")")
            self.print_not_reserved(node.left)

    def print_reserved(self):
        for node in self.reserved:
            print(node.name, ":", "(", node.start_interval,"->", node.end_interval,")")

    def print_tree(self):
        print("\n")
        print("Memoria reservada:")
        self.print_reserved()
        print("\n")
        print("Memoria no reservada:")
        self.print_not_reserved(self.root)
        print("\n")



def main(arg1):

    total_blocks = int(arg1)
    buddy_system = BuddySystem(total_blocks)

    while True:
        action = input("Ingresa una acción (RESERVAR, LIBERAR, MOSTRAR, SALIR): ").split()
        if action[0].lower() == "reservar":
            if len(action) < 3:
                print("Error: faltó información.")
                continue
            size = int(action[1])
            name = action[2]
            buddy_system.add_process(size, name)

        elif action[0].lower() == "liberar":
            if len(action) < 2:
                print("Error: faltó información.")
                continue
            name = action[1]
            buddy_system.deallocate_process(name)

        elif action[0].lower() == "mostrar":
            buddy_system.print_tree()

        elif action[0].lower() == "salir":
            break
        else:
            print("Error: Acción inválida.")

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("uso: python3 buddysystem.py <cantidad de bloques de memoria a manejar>")
        sys.exit(1)
    arg1 = sys.argv[1]
    main(arg1)

File: Parcial-I/Pregunta-III/buddysystem.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_LENGTH 50

struct Block {
    int start;
    int end;
    char name[MAX_LENGTH];
    struct Block* next;
};

struct Block* memory;

int is_power_of_two(int n) {
    if (n == 0) {
        printf("Debe reservar al menos un bloque de memoria.\n");
        exit(1);
    }

    return (n & (n - 1));
}

int which_power_of_two(int n) {
    if (n <= 0) {
        return -1; // No es una potencia de 2
    }
    int count = 0;
    while (n > 1) {
        if (n & 1) {
            return -1; // No es una potencia de 2
        }
        count++;
        n = n >> 1;
    }
    return count;
}

void initialize_memory(int total_blocks) {

    /* Si el número de bloques no es una
    potencia de 2, se obtiene la potencia
    de 2 más cercana.
    */
    while (is_power_of_two(total_blocks)) {
        total_blocks++;
    }

    int power_of_two = which_power_of_two(total_blocks);

    printf("power of two: %d\n", power_of_two);

    memory = (struct Block*)malloc(sizeof(struct Block));
    memory->start = 0;
    memory->end = total_blocks - 1;
    memory->next = NULL;
}

void reserve_memory(int size, char* name) {
    struct Block* current = memory;
    struct Block* previous = NULL;

    while (current != NULL) {
        int available_size = current->end - current->start + 1;

        if (available_size >= size) {
            if (current->next == NULL || current->next->start - current->end - 1 >= size) {
                struct Block* new_block = (struct Block*)malloc(sizeof(struct Block));
                new_block->start = current->end - size + 1;
                new_block->end = current->end;
                strcpy(new_block->name, name);
                new_block->next = current->next;
                current->end -= size;

                if (current->next != NULL) {
                    current->next->start -= size;
                }

                current->next = new_block;
                return;
            }
        }

        previous = current;
        current = current->next;
    }

    printf("No hay suficiente espacio libre para reservar %d bloques para %s.\n", size, name);
}

void release_memory(char* name) {
    // Implementar la liberación de memoria
}

void show_memory() {
    // Implementar la visualización de la memoria
}

int main(int argc, char* argv[]) {
    if (argc != 2) {
        printf("Por favor, proporcione la cantidad de bloques de memoria como argumento.\n");
        return 1;
    }

    int total_blocks = atoi(argv[1]);
    initialize_memory(total_blocks);

    char action[10], name[MAX_LENGTH];
    int size;

    while (1) {
        printf("Ingresa una acción (RESERVAR, LIBERAR, MOSTRAR, SALIR): ");
        scanf("%s", action);

        if (!strcmp(action, "RESERVAR")) {
            scanf("%d %s", &size, name);
            reserve_memory(size, name);
        } else if (!strcmp(action, "LIBERAR")) {
            scanf("%s", name);
            release_memory(name);
        } else if (!strcmp(action, "MOSTRAR")) {
            show_memory();
        } else if (!strcmp(action, "SALIR")) {
            break;
        } else {
            printf("Acción no reconocida. Inténtalo de nuevo.\n");
        }
    }

    return 0;
}
